hive3-metastore/src/test/java/org/apache/iceberg/hive/TestHiveClientPool.java

```diff
@@ -0,0 +1,225 @@
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.MockedStatic;
+import org.mockito.Mockito;
+
+public class TestHiveClientPool {
+
+  private static final String HIVE_SITE_CONTENT =
+      "<?xml version=\"1.0\"?>\n"
+          + "<?xml-stylesheet type=\"text/xsl\" href=\"configuration.xsl\"?>\n"
+          + "<configuration>\n"
+          + "  <property>\n"
+          + "    <name>hive.metastore.sasl.enabled</name>\n"
+          + "    <value>true</value>\n"
+          + "  </property>\n"
+          + "</configuration>\n";
+
+  HiveClientPool clients;
+
+  @BeforeEach
+  public void before() {
+    HiveClientPool clientPool = new HiveClientPool(2, new Configuration());
+    clients = Mockito.spy(clientPool);
+  }
+
+  @AfterEach
+  public void after() {
+    clients.close();
+    clients = null;
+  }
+
+  @Test
+  public void testConf() {
+    HiveConf conf = createHiveConf();
+    conf.set("hive.metastore.warehouse.dir", "file:/mywarehouse/");
+
+    HiveClientPool clientPool = new HiveClientPool(10, conf);
+    HiveConf clientConf = clientPool.hiveConf();
+
+    assertThat(clientConf.get("hive.metastore.warehouse.dir"))
+        .isEqualTo(conf.get("hive.metastore.warehouse.dir"));
+    assertThat(clientPool.poolSize()).isEqualTo(10);
+
+    // 'hive.metastore.sasl.enabled' should be 'true' as defined in xml
+    assertThat(clientConf.get("hive.metastore.sasl.enabled"))
+        .isEqualTo(conf.get("hive.metastore.sasl.enabled"));
+    assertThat(clientConf.getBoolVar(HiveConf.getConfVars("hive.metastore.sasl.enabled"))).isTrue();
+  }
+
+  private HiveConf createHiveConf() {
+    HiveConf hiveConf = new HiveConf();
+    try (InputStream inputStream =
+        new ByteArrayInputStream(HIVE_SITE_CONTENT.getBytes(StandardCharsets.UTF_8))) {
+      hiveConf.addResource(inputStream, "for_test");
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    return hiveConf;
+  }
+
+  @Test
+  public void testNewClientFailure() {
+    Mockito.doThrow(new RuntimeException("Connection exception")).when(clients).newClient();
+    assertThatThrownBy(() -> clients.run(Object::toString))
+        .isInstanceOf(RuntimeException.class)
+        .hasMessage("Connection exception");
+  }
+
+  @Test
+  public void testGetTablesFailsForNonReconnectableException() throws Exception {
+    HiveMetaStoreClient hmsClient = Mockito.mock(HiveMetaStoreClient.class);
+    Mockito.doReturn(hmsClient).when(clients).newClient();
+    Mockito.doThrow(new MetaException("Another meta exception"))
+        .when(hmsClient)
+        .getTables(Mockito.anyString(), Mockito.anyString());
+    assertThatThrownBy(() -> clients.run(client -> client.getTables("default", "t")))
+        .isInstanceOf(MetaException.class)
+        .hasMessage("Another meta exception");
+  }
+
+  @Test
+  public void testExceptionMessages() {
```

Comment:
@pvary this is the only test in `TestHiveClientPool` that is different, between Hive 2 on one hand and Hive 3 and 4 on the other.
The difference is due to `MetaStoreUtils.newInstance` (from Hive 2) no longer existing in Hive 3. In Hive 3 and 4 we have to use `JavaUtils.newInstance`.

Reply:
This shows how depending on the internals of the other libraries could cause issues later 😢 

I'm not too big fan of the methods created only for testing purposes, but maybe we can create a:
```
  @VisibleForTesting
  IMetaStoreClient newClientInternal() {
    return GET_CLIENT.invoke(
            hiveConf, (HiveMetaHookLoader) tbl -> null, HiveMetaStoreClient.class.getName());
  }
```

And then mock this out? This would save us quite a big amount of effort on other places.
Your thourgts?

Reply:
Thanks for the suggestion. I have implemented it.

Reply:
Looks much better 😄 
Thanks!